contracts/core/Eip7702Support.sol

```diff
@@ -0,0 +1,58 @@
+pragma solidity ^0.8;
+
+import "../interfaces/PackedUserOperation.sol";
+import "../core/UserOperationLib.sol";
+// SPDX-License-Identifier: MIT
+
+// EIP-7702 code prefix. Also, we use this prefix as a marker in the initCode. To specify this account is EIP-7702.
+uint256 constant EIP7702_PREFIX = 0xef0100;
+
+using UserOperationLib for PackedUserOperation;
+
+    //get alternate InitCode (just for hashing) when using EIP-7702
+    function _getEip7702InitCodeOverride(PackedUserOperation calldata userOp) view returns (bytes32) {
+        bytes calldata initCode = userOp.initCode;
+        if (! _isEip7702InitCode(initCode)) {
+            return 0;
+        }
+        address delegate = _getEip7702Delegate(userOp.getSender());
+        if (initCode.length < 20)
+            return keccak256(abi.encodePacked(delegate));
+        else
+            return keccak256(abi.encodePacked(delegate, initCode[20 :]));
+    }
+
+
+    function _isEip7702InitCode(bytes calldata initCode) pure returns (bool) {
+
+        if (initCode.length < 3) {
+            return false;
+        }
+        uint256 initCodeStart;
+        // solhint-disable-next-line no-inline-assembly
+        assembly ("memory-safe") {
+            initCodeStart := calldataload(initCode.offset)
+        }
+        // make sure first 20 bytes of initCode are "0xff0100" (padded with zeros)
+        // initCode can be shorter (e.g. only 3), but then it is already zero-padded.
+        return (initCodeStart >> (256 - 160)) == ((EIP7702_PREFIX << (160 - 24)));
+    }
+
+/**
+ * get the EIP-7702 delegate from contract code.
+ * requires EXTCODECOPY pr: https://github.com/ethereum/EIPs/pull/9248 (not yet merged or implemented)
+ **/
+    function _getEip7702Delegate(address sender) view returns (address) {
+        uint256 senderCode;
+
+        // solhint-disable-next-line no-inline-assembly
+        assembly ("memory-safe") {
+            extcodecopy(sender, 0, 0, 32)
+            senderCode := mload(0)
+        }
```

Comment:
You don't need any assembly, you can just do `sender.code`. Also if the code is not exactly 23 bytes long it should revert.

Reply:
> Even with sender.code, we'd need assembly to grab just the address out of it

I think Solidity now supports just converting `bytes` to `bytes32` to `uint256` so as far as I can tell there should be no difference. I may be wrong though.
```
        bytes memory b = sender.code;
        uint256(bytes32(b));
```

> No need to check it is 23 bytes since EIP-3541
I am concerned some up-and-coming L2s may be a little less rigorous in their implementation of EIP-3541 and that would allow some kind of an attack down the line. It would be dirt-cheap to just double-check.

Reply:
yes, it is possible. but it reads read entire code, and then extract just the first 32 bytes. 
rough check shows it to be 300 gas.
EIP-3541 doesn't really matter here: even if  account's code starts with "EF", it will never pass validation anyway (or any other execution call)

Reply:
> it reads read entire code, and then extract just the first 32 bytes

But the code is always 23 bytes long for EIP-7702, what is the problem here?
Please remove assembly if it is technically possible here. No need to overcomplicate things.

Reply:
I know it is 23 bytes and you know it too, but solidity doesn't... `sender.code` can possibly be any length, so it copies it as dynamic value before extracting the first word/